swift/trainers/utils.py

```diff
@@ -65,83 +60,6 @@ def is_instance_of_ms_model(model: Module) -> bool:
     return False
 
 
-def concat_template(feature: Dict, template: Template):
-    query: Optional[str] = feature.get('query', None)
-    system: Optional[str] = feature.get('system', None)
-    history: Optional[History] = feature.get('history', None)
-    if history is None:
-        history = []
-    if system is None:
-        if template.use_default_system:
-            system = template.default_system
-    else:
-        assert template.system_prefix is not None, 'not support `system`'
-    res_context_list: List[Context] = []
-    compute_loss_idx: List[float] = []
-    if system is None:
-        assert template.prefix != template.system_prefix, f'template.prefix: {template.prefix}'
-        prefix = template.prefix
-    else:
-        prefix = template.system_prefix
-    template._concat_context_list(prefix, res_context_list, compute_loss_idx, system=system)
-    for i, (q, r) in enumerate(history):
-        template._concat_context_list(
-            [
-                *template.prompt,
-                '{{RESPONSE}}',
-                *template.chat_sep  # noqa
-            ],
-            res_context_list,
-            compute_loss_idx,
-            query=q,
-            response=r,
-            round0=i)  # noqa
-    template._concat_context_list(template.prompt, res_context_list, compute_loss_idx, query=query, round0=len(history))
-    res_context_list, compute_loss_idx = template._simplify_context_list(res_context_list, compute_loss_idx)
-
-    return res_context_list, feature['response'], feature['rejected_response'], compute_loss_idx
-
-
-def sort_by_max_length(dataset: HfDataset, num_dataset: int, is_encoder_decoder: bool = False) -> HfDataset:
```

Comment:
Can be merged into `sort_by_max_length` func in llm/utils

Reply:
It hasn't been fully refactored yet. Once it's done, I might need to trouble you to review it.